services/web/client/source/class/qxapp/components/workbench/logger/LoggerView.js

```diff
@@ -0,0 +1,243 @@
+    },
+
+    info: function(who = "System", what = "") {
+      this.__addLog(who, what, LOG_LEVEL.info);
+    },
+
+    warn: function(who = "System", what = "") {
+      this.__addLog(who, what, LOG_LEVEL.warning);
+    },
+
+    error: function(who = "System", what = "") {
+      this.__addLog(who, what, LOG_LEVEL.error);
+    },
+
+    __addLog: function(who = "System", what = "", logLevel = 0) {
+      this.__logs.push({
+        who: who,
+        what: what,
+        logLevel: logLevel
+      });
+
+      const whoRich = this.__addWhoColorTag(who);
+      const whatRich = this.__addLevelColorTag(what, logLevel);
+      let msgLog = {
+        whoRich: whoRich,
+        whatRich: whatRich,
+        msg: {
+          who: who,
+          what: what,
+          logLevel: logLevel
+        }
+      };
+      this.__logModel.addRows([msgLog]);
+
+      this.__logModel.reloadData();
+    },
+
+    __addWhoColorTag: function(who) {
+      let whoColor = null;
+      for (let item of this.__messengerColors) {
+        if (item[0] === who) {
+          whoColor = item[1];
+          break;
+        }
+      }
+      if (whoColor === null) {
+        const luminanceBG = qxapp.utils.Utils.getColorLuminance(qxapp.theme.Color.colors["table-row-background-selected"]);
+        let luminanceText = null;
+        do {
+          whoColor = qxapp.utils.Utils.getRandomColor();
+          luminanceText = qxapp.utils.Utils.getColorLuminance(whoColor);
+        } while (Math.abs(luminanceBG-luminanceText) < 0.4);
+
+        this.__messengerColors.add([who, whoColor]);
+      }
+
+      return ("<font color=" + whoColor +">" + who + ": </font>");
+    },
+
+    __addLevelColorTag: function(what, logLevel) {
+      let logColor = null;
+      switch (logLevel) {
+        case LOG_LEVEL.debug:
+          logColor = qxapp.theme.Color.colors["logger-debug-message"];
+          break;
+        case LOG_LEVEL.info:
+          logColor = qxapp.theme.Color.colors["logger-info-message"];
+          break;
+        case LOG_LEVEL.warning:
+          logColor = qxapp.theme.Color.colors["logger-warning-message"];
+          break;
+        case LOG_LEVEL.error:
+          logColor = qxapp.theme.Color.colors["logger-error-message"];
+          break;
+        default:
+          logColor = qxapp.theme.Color.colors["logger-info-message"];
+          break;
+      }
+      return ("<font color=" + logColor +">" + what + "</font>");
+    },
+
+    __applyFilters: function() {
+      if (this.__logModel === null) {
+        return;
+      }
+
+      this.__logModel.setFilterString(this.__textfield.getValue());
+      this.__logModel.setFilterLogLevel(this.getLogLevel());
+      this.__logModel.reloadData();
+    },
+
+    __createInitMsg: function() {
+      const who = "System";
+      const what = "Logger intialized";
+      this.debug(who, what);
+    },
+
+    clearLogger: function() {
+      this.__logModel.clearTable();
+    }
+  }
+});
```

Comment:
there seem to be quite a lot of instances with 'do this 4 times' with only the log level changing in this code ...